### PR TITLE
fix: implement == / != for List, Set, Map, Result, and union types (#725)

### LINKED ISSUE #725: feat: implement == / != for collection and composite types (List, Map, Set, Result, union, closure)
## Summary

The `==` and `!=` operators are not supported for the following types, despite being useful for testing and program logic:

| Type | Error |
|------|-------|
| `List<T>` | `numeric operation requires numeric type, got pointer` |
| `Map<K,V>` | `numeric operation requires numeric type, got pointer` |
| `Set<T>` | `numeric operation requires numeric type, got pointer` |
| `closure` / function | `numeric operation requires numeric type, got pointer` |
| `Result<T,E>` | `numeric operation requires numeric type, got struct` |
| `union` (`T \| U`) | `numeric operation requires numeric type, got struct` |

Currently working: `int`, `float`, `str`, `bool`, `tuple`, `Option<T>`, `record`, `enum`.

## Reproducers

```ry
# List — error
[1, 2, 3] == [1, 2, 3]

# Map — error
m1: Map<str,int> = {"a": 1}
m2: Map<str,int> = {"a": 1}
m1 == m2

# Set — error
{1, 2} == {1, 2}

# Result — error
r1: Result<int, Error> = Ok(42)
r2: Result<int, Error> = Ok(42)
r1 == r2

# union — error
x: int | str = 42
y: int | str = 42
x == y
```

## Expected Behavior

- `List`: element-wise equality (`[1,2,3] == [1,2,3]` → `true`, `[1,2] == [1,3]` → `false`)
- `Map`: key-value pair equality
- `Set`: set membership equality (order-independent)
- `Result`: `Ok(v1) == Ok(v2)` iff `v1 == v2`; `Err(e1) == Err(e2)` iff messages equal; `Ok != Err`
- `union`: structural equality based on active variant and value
- `closure`: likely not needed (referential equality or unsupported is fine); document the decision

## Context

Discovered while implementing combinatorial test coverage in #628.

### PR DESCRIPTION
**Pull request by t0k0sh1**
$(cat <<'EOF'
## Summary

Implements `==` and `!=` for all previously unsupported composite types.

| Type | Approach |
|------|----------|
| `Result<T,E>` | Inline IR: compare `is_ok` flag, then inner `Ok`/`Err` value |
| `union A\|B` | `SwitchInst` per variant: compare tag, then inner value |
| `List<T>` | Inline IR loop: length check + element-wise comparison |
| `Set<T>` | Length check + `emitSubsetCheck` (order-independent) |
| `Map<K,V>` | Inline IR loop: length check + key lookup + value comparison |
| Closure | Compile-time error |

Supported element/value types for collections: `int`, `float`, `str`, `bool`.  
Complex element types (`List<record>`, nested collections) are tracked in #736.

Also includes the `#726` cherry-pick (Option inner value comparison) as a prerequisite.

## Test plan

- [ ] 33 new tests in `tests/spec/combinatorial/equality.test.ry` all pass
- [ ] `./build/ry_tests` passes (1425 tests)
- [ ] `./build/ry test -p` passes (90 files)
- [ ] ASan build passes
- [ ] `docs/reference/collections.md` and `docs/reference/types.md` updated

Closes #725
EOF
)

## Summary by CodeRabbit

* **New Features**
  * Added == / != support for List, Set, Map, Result, Option, tuples and union variants (closure equality remains unsupported).

* **Documentation**
  * Updated reference docs with explicit equality semantics and examples for collections, Result/Option, and union variants.

* **Tests**
  * Added ~113 combinatorial tests covering collections, equality, function arguments/returns, match behavior, nested types, syntax combinations, print/display, and stdlib boundary cases.

### CODE REVIEW

File: src/codegen_expr.cpp [hunk 1]
```diff
@@ -301,12 +301,50 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+        llvm::Value *bothNone = builder_.CreateAnd(
+            builder_.CreateNot(lhsFlag), builder_.CreateNot(rhsFlag), "both_none");
+        llvm::Value *bothSome = builder_.CreateAnd(lhsFlag, rhsFlag, "both_some");
+
+        llvm::Value *lhsInner = builder_.CreateExtractValue(lhs, 1, "opt_l_inner");
+        llvm::Value *rhsInner = builder_.CreateExtractValue(rhs, 1, "opt_r_inner");
+
+        llvm::Value *innerEq = emitComparisonOp("==", lhsInner, rhsInner, "", "");
+        llvm::Value *eqResult = builder_.CreateOr(
+            bothNone, builder_.CreateAnd(bothSome, innerEq), "opt_eq");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Fixed in the latest commit: Option equality now uses a conditional branch + PHI node so the inner comparison is only emitted when both operands are `Some`. The `None` path short-circuits without touching the undef inner value.

File: src/codegen_expr.cpp [hunk 2]
```diff
@@ -301,12 +301,50 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+        llvm::Value *errEq = emitComparisonOp("==",
+            builder_.CreateExtractValue(lhs, 2, "lhs_err_val"),
+            builder_.CreateExtractValue(rhs, 2, "rhs_err_val"), "", "");
+        // flagsEq && ((lhsOk && okEq) || (!lhsOk && errEq))
+        llvm::Value *innerEq = builder_.CreateOr(
+            builder_.CreateAnd(lhsOk, okEq, "ok_match"),
+            builder_.CreateAnd(builder_.CreateNot(lhsOk), errEq, "err_match"), "inner_eq");
+        llvm::Value *eqResult = builder_.CreateAnd(flagsEq, innerEq, "res_eq");
+        if (op == "!=") return builder_.CreateNot(eqResult, "res_ne");
+        return eqResult;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Fixed in the latest commit: Result equality now uses a 3-path control-flow structure (flags differ → false, both Ok → compare ok payloads, both Err → compare err payloads). Neither variant payload is compared when the flags differ.

File: src/codegen_expr.cpp [hunk 3]
```diff
@@ -337,9 +375,197 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+                        llvm::ConstantInt::get(llvm::cast<llvm::IntegerType>(i64Ty_), ci),
+                        caseBB);
+                    builder_.SetInsertPoint(caseBB);
+                    llvm::Type *compTy = uinfo.componentTypes[ci];
+                    llvm::Value *li = builder_.CreateLoad(compTy, lhsTmp, "ueq.li");
+                    llvm::Value *ri = builder_.CreateLoad(compTy, rhsTmp, "ueq.ri");
+                    llvm::Value *caseEq = emitComparisonOp("==", li, ri, "", "");
+                    phi->addIncoming(caseEq, builder_.GetInsertBlock());
+                    builder_.CreateBr(mergeBB);
+                }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:479]: Fixed in the latest commit: Union equality now checks all component type names against a primitive allowlist (`int`, `float`, `str`, `bool`) before emitting any comparison, and raises a codegen error for non-primitive variants.

File: src/codegen_expr.cpp [hunk 4]
```diff
@@ -337,9 +375,197 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+            llvm::Value *leqIv = builder_.CreateLoad(i64Ty_, leqI, "leq_iv");
+            builder_.CreateCondBr(builder_.CreateICmpSLT(leqIv, lf.len), bodyBB, mergeBB);
+            builder_.SetInsertPoint(bodyBB);
+            llvm::Value *leqIc = builder_.CreateLoad(i64Ty_, leqI, "leq_ic");
+            llvm::Value *le = builder_.CreateLoad(lhsElemTy,
+                builder_.CreateGEP(lhsElemTy, lf.data, {leqIc}, "leq_lep"), "leq_le");
+            llvm::Value *re = builder_.CreateLoad(lhsElemTy,
+                builder_.CreateGEP(lhsElemTy, rf.data, {leqIc}, "leq_rep"), "leq_re");
+            builder_.CreateCondBr(emitComparisonOp("==", le, re, "", ""), nextBB, failBB);
+            builder_.SetInsertPoint(nextBB);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:527]: Fixed in the latest commit: List equality now rejects pointer element types that are not plain strings — specifically, it raises a codegen error for `List<List<T>>` (detected via `NestedListElem` type metadata) to prevent metadata-loss misclassification.

File: src/codegen_expr.cpp [hunk 5]
```diff
@@ -337,9 +375,197 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+            builder_.CreateCondBr(
+                builder_.CreateICmpSGE(rhsIdx, llvm::ConstantInt::get(i64Ty_, 0), "meq_found"),
+                valBB, failBB);
+            builder_.SetInsertPoint(valBB);
+            llvm::Value *lv = builder_.CreateLoad(valTy,
+                builder_.CreateGEP(valTy, lf.vals, {meqIc}, "meq_lvep"), "meq_lv");
+            llvm::Value *rv = builder_.CreateLoad(valTy,
+                builder_.CreateGEP(valTy, rf.vals, {rhsIdx}, "meq_rvep"), "meq_rv");
+            builder_.CreateCondBr(emitComparisonOp("==", lv, rv, "", ""), nextBB, failBB);
+            builder_.SetInsertPoint(nextBB);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:631]: Fixed in the latest commit: Map equality now rejects maps whose value type is a non-string pointer (detected via `map_value_type_name` being non-empty) and raises a codegen error, preventing metadata-loss misclassification.

File: src/codegen_expr.cpp [hunk 6]
```diff
@@ -337,9 +375,197 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+        }
+    }
+
+    // List equality: element-wise comparison (only == and != supported)
+    {
+        llvm::Type *lhsElemTy = getListElementType(lhs);
+        llvm::Type *rhsElemTy = getListElementType(rhs);
+        if (lhsElemTy && rhsElemTy && (op == "==" || op == "!=")) {
+            if (lhsElemTy != rhsElemTy)
+                codegenError("cannot compare List values with different element types");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:494]: Fixed in the latest commit: List equality now guards against non-string pointer element types (e.g. `List<List<T>>`) by checking for `NestedListElem` type metadata, and similarly rejects `Map` value pointer types via `map_value_type_name`. These cases raise a codegen error instead of silently misclassifying the loaded pointers as strings.

File: src/jit_runner.cpp [hunk 7]
```diff
@@ -355,6 +359,20 @@ int runRySource(const std::string &src, const std::string &source_name,
...
+        llvm::consumeError(std::move(err));
+
+    // Intentionally leak the LLJIT object to prevent a crash in ~LLJIT() on
+    // Linux ELF.  The crash occurs in ~ExecutorProcessControl() when freeing
+    // LLVM-internal heap structures (SymbolStringPool) that are corrupted by
+    // JIT relocation side-effects specific to ELF+JITLink.  For a short-lived
+    // CLI subprocess this leak is acceptable; the OS reclaims memory on exit.
+    // TODO(#742): Investigate root cause; fix or file upstream LLVM bug.
+    jit.release();
+
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/jit_runner.cpp]: The leak is now gated to Linux only (`#ifndef __APPLE__`) and is bounded by the process lifetime — `ry` always exits after a run/test command. The root cause of the `~LLJIT()` crash is tracked in #742 for further investigation. We considered a non-leaking workaround but could not find one that avoids the crash on Linux ELF+JITLink; the TODO comment reflects this.

File: tests/spec/combinatorial/stdlib_boundary.test.ry [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    expect(parts[1]).to_eq("い")
+    expect(parts[2]).to_eq("う")
+  )
+  it("should get char_at for unicode string with negative index", ():
+    expect(char_at("あいう", -1)).to_eq("う")
+    expect(char_at("あいう", -3)).to_eq("あ")
+  )
+  it("should compute substring for unicode string", ():
+    s = "abcdef"
+    expect(substring(s, 1, 4)).to_eq("bcd")
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/combinatorial/stdlib_boundary.test.ry]: Fixed in the latest commit: the test now uses `"あいうえお"` as input and expects `"いうえ"` as output, exercising actual multi-byte Unicode characters in the boundary test.

File: src/jit_runner.cpp [hunk 9]
```diff
@@ -355,6 +359,11 @@ int runRySource(const std::string &src, const std::string &source_name,
         cs->file_id_offset += fc;
     }
 
+    // Explicitly remove JIT resources before ~LLJIT() to avoid Linux ELF
+    // teardown crash.  Errors here are non-fatal (program already ran).
+    if (auto err = RT->remove())
+        llvm::consumeError(std::move(err));
+
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/jit_runner.cpp]: Fixed in the latest commit: `#include <llvm/ADT/ScopeExit.h>` has been added and `llvm::make_scope_exit` is used for the RAII cleanup guard.

File: src/codegen_expr.cpp [hunk 10]
```diff
@@ -301,12 +302,85 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
         llvm::Value *lhsFlag = builder_.CreateExtractValue(lhs, 0, "lhs_has");
         llvm::Value *rhsFlag = builder_.CreateExtractValue(rhs, 0, "rhs_has");
-        if (op == "==") return builder_.CreateICmpEQ(lhsFlag, rhsFlag, "opt_eq");
-        return builder_.CreateICmpNE(lhsFlag, rhsFlag, "opt_ne");
+
+        // When inner types differ (e.g., Option<Error> vs none's Option<int>),
+        // only compare has_value flags (one side must be None)
+        if (lhs->getType() != rhs->getType()) {
+            if (op == "==") return builder_.CreateICmpEQ(lhsFlag, rhsFlag, "opt_eq");
+            return builder_.CreateICmpNE(lhsFlag, rhsFlag, "opt_ne");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: When `lhs->getType() != rhs->getType()`, one operand is always a `none` literal whose LLVM type was not widened to the other side (e.g. `Error? == none`). The Ry type checker rejects `Some(a)==Some(b)` with incompatible inner types before codegen is reached. Added a comment to the code to document this invariant.

File: src/codegen_expr.cpp [hunk 11]
```diff
@@ -337,9 +411,244 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+        llvm::Type *rhsElemTy = getListElementType(rhs);
+        if (lhsElemTy && rhsElemTy && (op == "==" || op == "!=")) {
+            if (lhsElemTy != rhsElemTy)
+                codegenError("cannot compare List values with different element types");
+            // Reject equality for non-string pointer elements.
+            // Loaded list elements have no ValueMetadata; if lhsElemTy == ptrTy_
+            // but elements are not strings (e.g. List<List<T>>), isStringValue()
+            // would misclassify them and emit strcmp on arbitrary pointers.
+            if (lhsElemTy == ptrTy_ && getTypeMeta(TypeMeta::NestedListElem, lhs))
+                codegenError("list == / != is not supported for element type 'List<T>'");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:500]: Deferred to #727 (`fix: type info lost when accessing elements of List<Map>, List<Set>, List<closure>`), where `list_elem_type_name` metadata will be added. Once that metadata is available, the equality guard can be extended to also reject `List<Map>`, `List<Set>`, and `List<closure>` elements. Tracking note added to #727.

File: src/codegen_expr.cpp [hunk 12]
```diff
@@ -337,9 +411,244 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+                llvm::BasicBlock *entryBB = builder_.GetInsertBlock();
+                builder_.CreateCondBr(
+                    builder_.CreateICmpEQ(lhsTag, rhsTag, "utag_eq"), sameTagBB, mergeBB);
+
+                builder_.SetInsertPoint(sameTagBB);
+                auto *dataTy = uinfo.llvmType->getElementType(1);
+                llvm::AllocaInst *lhsTmp = builder_.CreateAlloca(dataTy, nullptr, "ueq.ld");
+                llvm::AllocaInst *rhsTmp = builder_.CreateAlloca(dataTy, nullptr, "ueq.rd");
+                lhsTmp->setAlignment(mod_->getDataLayout().getABITypeAlign(uinfo.llvmType));
+                rhsTmp->setAlignment(mod_->getDataLayout().getABITypeAlign(uinfo.llvmType));
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Fixed: alloca alignment now uses `dataTy` (the payload type) instead of `uinfo.llvmType` (the whole union struct). Using the payload type alignment is more precise and correct for the alloca being created.

File: src/codegen_expr.cpp [hunk 13]
```diff
@@ -337,9 +411,244 @@ llvm::Value *CodeGen::emitComparisonOp(const std::string &op, llvm::Value *lhs,
...
+            // Short-circuit: skip the subset-check loop when lengths differ.
+            llvm::Function   *curFn    = builder_.GetInsertBlock()->getParent();
+            llvm::BasicBlock *startBB  = builder_.GetInsertBlock();
+            llvm::BasicBlock *checkBB  = llvm::BasicBlock::Create(*ctx_, "seq.check", curFn);
+            llvm::BasicBlock *mergeBB  = llvm::BasicBlock::Create(*ctx_, "seq.merge",  curFn);
+
+            builder_.CreateCondBr(lenEq, checkBB, mergeBB);
+
+            builder_.SetInsertPoint(checkBB);
+            llvm::Value *isSubset = emitSubsetCheck(lhs, rhs, "seq_sub");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:563]: Deferred to #727, same as the List<Map/Set/closure> issue above. Once `list_elem_type_name` and analogous Set element metadata are available, the Set equality guard can be extended to reject non-string pointer elements. Tracking note added to #727.

File: src/jit_runner.cpp [hunk 14]
```diff
@@ -355,6 +367,17 @@ int runRySource(const std::string &src, const std::string &source_name,
...
+#ifndef __APPLE__
+    // Intentionally leak the LLJIT on Linux to prevent a crash in ~LLJIT()
+    // (~ExecutorProcessControl() → __libc_free) caused by JIT relocation
+    // side-effects on ELF+JITLink.  The OS reclaims memory on process exit.
+    // Affects both subprocess (ry test -p) and sequential (ry test) modes;
+    // since ry always exits after the run/test command completes, the
+    // per-invocation leak is bounded by the process lifetime.
+    // TODO(#742): Investigate root cause; fix or file upstream LLVM bug.
+    jit.release();
+#endif
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/jit_runner.cpp]: Fixed: changed `#ifndef __APPLE__` to `#ifdef __linux__` — the crash is ELF+JITLink specific so the workaround should be Linux-only, not "everything except Apple".
